rllib/core/models/base.py

```diff
@@ -0,0 +1,380 @@
+from ray.rllib.models.specs.specs_base import Spec
+from ray.rllib.models.specs.specs_dict import SpecDict
+from ray.rllib.utils.annotations import ExperimentalAPI
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.nested_dict import NestedDict
+from ray.rllib.models.specs.checker import convert_to_canonical_format
+from ray.rllib.utils.typing import TensorType
+
+# Top level keys that unify model i/o.
+STATE_IN: str = "state_in"
+STATE_OUT: str = "state_out"
+ENCODER_OUT: str = "encoder_out"
+# For Actor-Critic algorithms, these signify data related to the actor and critic
+ACTOR: str = "actor"
+CRITIC: str = "critic"
+
+
+@ExperimentalAPI
+@dataclass
+class ModelConfig(abc.ABC):
+    """Base class for model configurations.
+
+    ModelConfigs are framework-agnostic.
+    A ModelConfig is usually built by RLModules after getting it from a Catalog object.
+    It is therefore a means of configuration for RLModules. However, ModelConfigs are
+    not restricted to be used only with Catalog or RLModules.
+    A usage Example together with a Model can be found in the Model.
+
+    Attributes:
+        output_dim: The output dimension of the network.
+    """
+
+    output_dim: int = None
+
+    @abc.abstractmethod
+    def build(self, framework: str):
+        """Builds the model.
+
+        Args:
+            framework: The framework to use for building the model.
+        """
+        raise NotImplementedError
+
+
+class Model(abc.ABC):
+    """Framework-agnostic base class for RLlib models.
+
+    Models are low-level neural network components that offer input- and
+    output-specification, a forward method, and a get_initial_state method. Models
+    are composed in RLModules.
+
+    Usage Example together with ModelConfig:
+
+    .. testcode::
+
+        from ray.rllib.core.models.base import Model, ModelConfig
+        from dataclasses import dataclass
+
+        class MyModel(Model):
+            def __init__(self, config):
+                super().__init__(config)
+                self.my_param = config.my_param * 2
+
+            def _forward(self, input_dict):
+                return input_dict["obs"] * self.my_param
+
+
+        @dataclass
+        class MyModelConfig(ModelConfig):
+            my_param: int = 42
+
+            def build(self, framework: str):
+                if framework == "bork":
+                    return MyModel(self)
+
+
+        config = MyModelConfig(my_param=3)
+        model = config.build(framework="bork")
+        print(model._forward({"obs": 1}))
+
+    .. testoutput::
+
+        6
+
+    """
+
+    def __init__(self, config: ModelConfig):
+        self.config = config
+
+        # These should not be overridden by subclasses. Instead, get_input_spec and
+        # get_output_spec should be overridden.
+        self._input_spec = self.get_input_spec()
+        self._output_spec = self.get_output_spec()
+
+    def get_input_spec(self) -> Union[Spec, None]:
```

Comment:
nit (do only if you gotta make changes because of some other reasons): decorate with overrideToMakeCustomization flag

Reply:
Damn I was looking for that one and could not find it.

Reply:
use source graph man :) https://sourcegraph.com/github.com/ray-project/ray/-/blob/rllib/utils/annotations.py?L110:5&subtree=true